certora/specs/VaultERC4626.spec

```diff
@@ -0,0 +1,573 @@
+    init_state axiom sumOfBalances == 0;
+}
+
+hook Sstore currentContract.vaultStorage.users[KEY address addy].data Vault.PackedUserSlot newValue (Vault.PackedUserSlot oldValue)  {
+    sumOfBalances = sumOfBalances + newValue - oldValue;
+}
+
+hook Sload Vault.PackedUserSlot val currentContract.vaultStorage.users[KEY address addy].data  {
+    require sumOfBalances >= to_mathint(val);
+}
+
+// passing: https://prover.certora.com/output/65266/de3636d287d2473294463c07263fc11e/?anonymousKey=ac8f74e6c5c1298f0954a21fafd41cccf32b9ffb
+invariant totalSupplyIsSumOfBalances(env e)
+    // to_mathint(totalSupply(e)) == sumOfBalances + accumulatedFees(e);
+    to_mathint(totalSupply(e)) == sumOfBalances;
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+////                    #     State Transition                             /////
+////////////////////////////////////////////////////////////////////////////////
+
+//run: https://prover.certora.com/output/65266/3ef25c98a7e34422bcf177d853662b5f/?anonymousKey=ca43e967a607a404f34b39c70f6517e90dac0902
+rule totalsMonotonicity() {
+    method f; env e; calldataarg args;
+    require e.msg.sender != currentContract; 
+    uint256 totalSupplyBefore = totalSupply(e);
+    uint256 totalAssetsBefore = totalAssets(e);
+    address receiver;
+    safeAssumptions(e, receiver, e.msg.sender);
+    callReceiverFunctions(f, e, receiver);
+
+    uint256 totalSupplyAfter = totalSupply(e);
+    uint256 totalAssetsAfter = totalAssets(e);
+    
+    // possibly assert totalSupply and totalAssets must not change in opposite directions
+    assert totalSupplyBefore < totalSupplyAfter  <=> totalAssetsBefore < totalAssetsAfter,
+        "if totalSupply changes by a larger amount, the corresponding change in totalAssets must remain the same or grow";
+    assert totalSupplyAfter == totalSupplyBefore => totalAssetsBefore == totalAssetsAfter,
+        "equal size changes to totalSupply must yield equal size changes to totalAssets";
+}
+
+rule underlyingCannotChange() {
+    address originalAsset = asset();
+
+    method f; env e; calldataarg args;
+    f(e, args);
+
+    address newAsset = asset();
+
+    assert originalAsset == newAsset,
+        "the underlying asset of a contract must not change";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+////                    #   High Level                                    /////
+////////////////////////////////////////////////////////////////////////////////
+
+// passing
+// run: https://prover.certora.com/output/65266/1912c053cdf8485087f2c050146c64aa/?anonymousKey=a12e3d573258a4d8136a19b612448a50f80b9a21
+rule dustFavorsTheHouse(uint assetsIn )
+{
+    env e;
+        
+    require e.msg.sender != currentContract;
+    safeAssumptions(e,e.msg.sender,e.msg.sender);
+    uint256 totalSupplyBefore = totalSupply(e);
+
+    uint balanceBefore = currentContract.balanceOf(e, currentContract);
```

Comment:
shouldn't it be `uint balanceBefore = IERC20(asset()).balanceOf(e, currentContract);`?

Reply:
You are right. I fixed this now (note that the balance you suggested checking is the one returned by our harness function `ERC4626Harness.userAssets(currentContract)`). It passes with this fix: https://prover.certora.com/output/65266/a19010e64bb8424aa513be8b75d15cdf/?anonymousKey=87c73cdf676930336269396f2dbb3cac3d78b997